### PR TITLE
[PLAT-6001] BSGEventUploader

### PR DESCRIPTION
**Pull request by nickdowell**
## Goal

Implement a new delivery system that will be able to work with events that are not saved as KSCrashReports.

Part of the App Hangs project.

## Design

Follows Engineering Design proposal.

## Changeset

* `BSGEventUploader` implements the core logic of scanning stored events and managing the NSOperationQueue.
* `BSGEventUploadOperation` implements the core logic of processing the event, preparing and sending the request. Subclasses of this implement behaviour that differs between source type.
* `AutoSessionMixedEventsScenario` required reworking because there is no longer a 1 second delay before sending starts - this scenario was ending up with intermittent duplicates being received as a consequence.

## Testing

* Tested locally using example app(s)
* Tested via E2E scenarios

### CODE REVIEW

**Review by kstenerud — CHANGES_REQUESTED** [review]

File: Bugsnag/Delivery/BSGEventUploadFileOperation.m [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+- (instancetype)initWithFile:(NSString *)file delegate:(id<BSGEventUploadOperationDelegate>)delegate {
+    if (self = [super initWithDelegate:delegate]) {
+        _file = [file copy];
+    }
+    return self;
+}
+
+- (BugsnagEvent *)loadEvent {
+    // TODO: Load a BugsnagEvent from json file on disk
```
[the remark this answers is not in the document]
Comment (nickdowell) [reply Bugsnag/Delivery/BSGEventUploadFileOperation.m:26]: Yes - upload of bugsnag-formatted events from disk is planned to be part of a future PR

File: Bugsnag/Delivery/BSGEventUploadKSCrashReportOperation.m [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+@implementation BSGEventUploadKSCrashReportOperation
+
+- (BugsnagEvent *)loadEventAndReturnError:(NSError **)errorPtr {
+    NSData *data = [NSData dataWithContentsOfFile:self.file options:0 error:errorPtr];
+    if (!data) {
+        return nil;
+    }
+    
+    id json = [BSGJSONSerialization JSONObjectWithData:data options:0 error:errorPtr];
```
[the remark this answers is not in the document]
Comment (nickdowell) [reply Bugsnag/Delivery/BSGEventUploadKSCrashReportOperation.m:30]: Correct; if the JSON is invalid then the report will not be sent. This is intended & same as the current behaviour.

File: Bugsnag/Delivery/BSGEventUploadKSCrashReportOperation.m [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+    [self mergeDictWithKey:@BSG_KSCrashField_UserAtCrash
+           intoDictWithKey:@BSG_KSCrashField_User
+                  inReport:mutableReport];
+
+    NSMutableDictionary *crashReport =
+            [report[@BSG_KSCrashField_Crash] mutableCopy];
+    mutableReport[@BSG_KSCrashField_Crash] = crashReport;
+    BSG_KSCrashDoctor *doctor = [BSG_KSCrashDoctor doctor];
+    crashReport[@BSG_KSCrashField_Diagnosis] = [doctor diagnoseCrash:report];
```
[the remark this answers is not in the document]
Comment (nickdowell) [reply Bugsnag/Delivery/BSGEventUploadKSCrashReportOperation.m:79]: Yes; it is used by [BSGParseErrorMessage()](https://github.com/bugsnag/bugsnag-cocoa/blob/nickdowell/event-uploader/Bugsnag/Payload/BugsnagError.m#L65-L73)

File: Bugsnag/Delivery/BSGEventUploadObjectOperation.m [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+//  Copyright © 2021 Bugsnag Inc. All rights reserved.
+//
+
+#import "BSGEventUploadObjectOperation.h"
+
+#import "BSGJSONSerialization.h"
+#import "BugsnagEvent.h"
+#import "BugsnagLogger.h"
+
+@implementation BSGEventUploadObjectOperation
```
[the remark this answers is not in the document]
Comment (nickdowell) [reply Bugsnag/Delivery/BSGEventUploadObjectOperation.m:15]: The purpose of this class is to upload a `BugsnagEvent` that has not been persisted to disk.

The event gets persisted to disk if the upload fails, and its superclass contains the logic for uploading.

File: Bugsnag/Delivery/BSGEventUploadOperation.m [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+// MARK: Subclassing
+
+- (BugsnagEvent *)loadEventAndReturnError:(NSError **)errorPtr {
+    return nil;
+}
+
+- (void)deleteEvent {
+}
+
+- (void)storeEventPayload:(NSDictionary *)eventPayload inDirectory:(NSString *)directory {
```
[the remark this answers is not in the document]
Comment (nickdowell) [reply Bugsnag/Delivery/BSGEventUploadOperation.m:139]: I think this is probably worth discussing in person, it's a consequence of the design choice to use subclassing to represent the different kinds of uploads.

[the remark this answers is not in the document]
Comment (kstenerud) [reply Bugsnag/Delivery/BSGEventUploadOperation.m:139]: After in-person discussion, I have no issues with this

**Review by kstenerud — APPROVED** [review]

**Comment by nickdowell** [comment]
This PR was merged manually to resolve the conflicts

**Pull request closed without merging**